addons/pos_pinelabs/static/src/app/utils/payment_pinelabs.js

```diff
@@ -0,0 +1,204 @@
+import { _t } from "@web/core/l10n/translation";
+import { PaymentInterface } from "@point_of_sale/app/payment/payment_interface";
+import { AlertDialog } from "@web/core/confirmation_dialog/confirmation_dialog";
+import { serializeDateTime } from "@web/core/l10n/dates";
+import { registry } from "@web/core/registry";
+
+const CANCEL_REQUEST_TIME_LIMIT = 125000;
+const { DateTime } = luxon;
+
+export class PaymentPinelabs extends PaymentInterface {
+    setup() {
+        super.setup(...arguments);
+        this.paymentInitialTime = null;
+        this.paymentLineResolvers = {};
+        this.offlineErrorHandler = registry.category("error_handlers").get("offlineErrorHandler");
+    }
+
+    send_payment_request(cid) {
+        super.send_payment_request(cid);
+        return this._processPinelabs();
+    }
+
+    pendingPinelabsPaymentLine() {
+        return this.pos.getPendingPaymentLine("pinelabs");
+    }
+
+    send_payment_cancel(order, cid) {
+        super.send_payment_cancel(order, cid);
+        return this._pinelabsCancel();
+    }
+
+    _callPinelabs(data, action) {
+        return this.pos.data
+            .call("pos.payment.method", action, [[this.payment_method_id.id], data])
+            .catch((error) => {
+                const line = this.pendingPinelabsPaymentLine();
+                if (line) {
+                    line.set_payment_status("retry");
+                }
+                this.offlineErrorHandler(this.env, error, error);
+            });
+    }
+
+    /**
+     * This method processes the response from Pinelabs
+     * when we request payment or cancellation.
+     */
+    async _pinelabsHandleResponse(response) {
+        const line = this.pendingPinelabsPaymentLine();
+        if (!response || response.error) {
+            line.set_payment_status("retry");
+            response.error ? this._showError(response.error) : null;
+            return false;
+        }
+        line.set_payment_status("waitingCard");
+        line.plutus_transaction_reference = response.plutusTransactionReferenceID;
+        return await this._waitForPaymentToConfirm();
+    }
+
+    async _pinelabsCancel() {
```

Comment:
show the PTR number in the notification to reduce confusion from the cashier side.